Refactor ServiceProvider and related code

monfresh opened the pull request:
**Why**: To simplify it and remove unused code. The main purpose of
this class is to provide a hash of service provider attributes based
on the contents of the `service_providers.yml` file. This means that
all we need to do is parse the file, symbolize the keys, process some
of the values, such as the cert, and generate a fingerprint from the
cert.

The only instance methods needed are `metadata`, which returns the
hash of attributes, and `encryption opts`. All the instance methods
based on the YAML keys that we used to have are not needed. They were
never called (or didn't need to be called) directly. I'm not sure what
I was thinking when I wrote that code. Note that `encryption_opts`
could be merged into the `metadata` hash as well, but it's much easier
to test when it's a separate method.

**How**:
- Remove all instance methods based on YAML file keys.
- To return the proper hash in the `metadata` method, parse the YAML
file, symbolize the keys, and merge any other attributes that aren't
read from the YAML file, such as the fingerprint.
- Make the `block_encryption` key required in the YAML file as opposed
to using a conditional in the code.
- Remove `metadata_url` entries from the YAML file since those are not
used. None of those URLs exist.
- Pass in the `OpenSSL::X509::Certificate` instance of the cert to the
Fingerprinter class, as opposed to creating two instances, one in the
ServiceProvider class, and one in the Fingerprinter class.
- Raise an error if the issuer is invalid (i.e not present in the YAML
file).

monfresh commented:
Please merge #427 first, and then I'll rebase to allow Travis to pass.

pkarman commented:
lgtm

pkarman commented:
@monfresh this doesn't appear to need a re-base. ok to merge?

monfresh commented:
Thumbs up from me, but I'd like @sbc100 to take a look as well.

sbc100 commented on app/services/service_provider_config.rb:
```diff
@@ -0,0 +1,22 @@
+class ServiceProviderConfig
+  def initialize(filename:, issuer:)
+    @issuer = issuer
+    @data ||= YAML.load_file("#{Rails.root}/config/#{filename}")
```
The `||=` looks pointless here?  Presumably you are trying to load the yaml file just one but does this actually do that?

monfresh replied:
Looks like it's not needed.

sbc100 commented on app/services/service_provider_config.rb:
```diff
@@ -0,0 +1,22 @@
+class ServiceProviderConfig
+  def initialize(filename:, issuer:)
```
What do the trailing colons do here?

monfresh replied:
That is Ruby 2.0's keyword arguments. The colon means the arguments are required, and to use this method, you have to name the arguments, which makes it easy to understand what each arg represents. For example, instead of calling a method like this:
```ruby
method_with_vague_args('service_providers.yml', 'foo')
```
you do this:
```ruby
method_with_clear_args(filename: 'service_providers.yml', issuer: 'test.host')
```
More info here:
https://robots.thoughtbot.com/ruby-2-keyword-arguments

sbc100 replied:
Why are you using this syntax here in particular?  Should we start using it everywhere?  What are the rules for when we should or shouldn't use this?

monfresh replied:
I'm using it here for clarity. There are no hard and fast rules, but we can certainly come up with some. I would say use it when you think there might be confusion about what an argument represents.

sbc100 commented on lib/service_provider.rb:
```diff
@@ -1,105 +1,65 @@
-  def sp_attributes
-    config[:valid_hosts].fetch(@issuer, {})
+  def fingerprint
+    @fingerprint ||= ::Fingerprinter.fingerprint_cert(ssl_cert)
```
Whats does the `::` do here?

monfresh replied:
Looks like it's not needed. It was already there so I didn't change it. The tests pass without the colons so I'll remove them. Presumably at one point there was an issue with accessing without the colons, probably due to a missing `require` statement somewhere, and due to Fingerprinter being inside the `lib` directory, which doesn't automatically get required by Rails.

sbc100 commented on spec/controllers/saml_idp_controller_spec.rb:
```diff
@@ -203,23 +197,22 @@
-        expect(response.status).to eq(403)
+      it 'raises an invalid service provider error' do
+        expect { head '/api/saml/auth?SAMLRequest=bang!' }.
+          to raise_error 'nil is an invalid service provider!'
```
Shouldn't we handle this and return an http error?  My understanding is that when assertions show up at runtime like this that points to bug in the code (at least thats how think about it in python world).

monfresh replied:
You're right. An exception is not the right thing here. I'll revert it to an HTTP 403 error.

sbc100 commented on spec/lib/fingerprinter_spec.rb:
```diff
@@ -0,0 +1,45 @@
+
+    context 'ssl_cert is present' do
+      it 'returns a hexdigest of the cert' do
+        def cert
```
Why not just `cert = \`?

monfresh replied:
No particular reason. I can change this.

monfresh commented:
@sbc100 I addressed your comments. I will handle the invalid service provider scenario in a new PR since that is new functionality. We don't currently handle that scenario.

sbc100 commented on app/services/service_provider_config.rb:
```diff
@@ -0,0 +1,22 @@
+class ServiceProviderConfig
+  def initialize(filename:, issuer:)
+    @issuer = issuer
+    @data = YAML.load_file("#{Rails.root}/config/#{filename}")
```
Does this code run more than once per request?   Even if its only once per request wouldn't it be better not to load and parse this for every request?

monfresh replied:
Good point. I'll load it once in an initializer.

sbc100 commented on spec/services/attribute_asserter_spec.rb:
```diff
@@ -25,9 +31,8 @@
-            %w(email phone first_name)
-          )
+          allow(service_provider.metadata).to receive(:[]).with(:attribute_bundle).
+            and_return(%w(email phone first_name))
```
What about just `service_provider.metadata['attribute_bundle'] = %w(email phone first_name)`

monfresh replied:
The metadata hash is not modifiable. It's read-only.

sbc100 commented on app/services/service_provider_config.rb:
```diff
@@ -0,0 +1,22 @@
+      data.merge!(data.fetch('superb.legit.domain.gov', {}))
+    else
+      data.merge!(data.fetch(Rails.env, {}))
+    end
```
Do this once in the constructor rather than every time `data_hash` is called?

monfresh replied:
This will be part of the initializer, so it will only run once.

sbc100 commented on lib/service_provider.rb:
```diff
@@ -1,105 +1,63 @@
-
-  def metadata_url
-    sp_attributes['metadata_url']
+    sp_attributes.merge(shared_attributes)
```
This is creating a new hash every time its called, right?  Why not merge! once?

monfresh replied:
Good catch! `merge! ` is 24x faster than `merge`: https://github.com/JuanitoFatas/fast-ruby#hashmerge-vs-hashmerge-code

sbc100 commented on lib/service_provider.rb:
```diff
@@ -1,105 +1,63 @@
 
-  def encryption_opts
-    return nil unless encrypt_responses?
+  def shared_attributes
```
What do you mean my shared attributed here?   Its seems like maybe it should be `added_attributes` or `synthetic_attributes`, maybe?

monfresh replied:
I'm getting rid of the method altogether and just merging the fingerprint directly into `sp_attributes`. No need to use a separate method and try to come up with a good name when there's only one attribute to add.

monfresh commented:
@sbc100 I addressed your comments in a second commit. PTAL.

sbc100 commented on config/initializers/service_providers.rb:
```diff
@@ -0,0 +1,3 @@
+SERVICE_PROVIDERS = YAML.load_file("#{Rails.root}/config/service_providers.yml")
```
I'm wondering why do this here rather than at the ServiceProviderConfig class level.   e.g.. why no just have `@@service_proveriders ||=` in the initialiser?   Seems like it would be nicer since it would keep all the code together.   I guess the good thing about static inits is maybe that they push the startup cost to startup time and errors show at startup?   However I'm not sure the cost here is high enough to warrant a special initializer file?

monfresh replied:
Good point. I didn't think to use a class variable, but this is a great use for one. I'll fix and push again.

monfresh replied:
Actually, that's not working. In order to test that the proper config is loaded based on the domain name, I need to be able to call a method to update the hash, since it's memoized. I wasn't able to get it to work with class variables. I'll merge this as is, and we can revisit later if there's a way to do it from the same class.

sbc100 commented:
lgtm % comment.  If you like the current init style i'm happy for your to merge this.
